azure/scope/machinepool.go

```diff
@@ -158,61 +163,216 @@ func (m *MachinePoolScope) ProvisioningState() infrav1.ProvisioningState {
-// is running on the node, the provider IDs for the instances and the providerIDList for the AzureMachinePool spec.
-func (m *MachinePoolScope) UpdateInstanceStatuses(ctx context.Context, instances []infrav1exp.VMSSVM) error {
-	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolScope.UpdateInstanceStatuses")
-	defer span.End()
+// NeedsRequeue return true if any machines are not on the latest model or the VMSS is not in a terminal provisioning
+// state.
+func (m *MachinePoolScope) NeedsRequeue() bool {
+	state := m.AzureMachinePool.Status.ProvisioningState
+	if m.vmssState == nil {
+		return state != nil && infrav1.IsTerminalVMState(*state)
+	}
 
-	providerIDs := make([]string, len(instances))
-	for i, instance := range instances {
-		providerIDs[i] = azure.ProviderIDPrefix + instance.ID
+	if !m.vmssState.HasLatestModelAppliedToAll() {
+		return true
 	}
 
-	nodeStatusByProviderID, err := m.getNodeStatusByProviderID(ctx, providerIDs)
+	desiredMatchesActual := len(m.vmssState.Instances) == int(m.DesiredReplicas())
+	return !(state != nil && infrav1.IsTerminalVMState(*state) && desiredMatchesActual)
+}
+
+// DesiredReplicas returns the replica count on machine pool or 0 if machine pool replicas is nil.
+func (m MachinePoolScope) DesiredReplicas() int32 {
+	return to.Int32(m.MachinePool.Spec.Replicas)
+}
+
+// MaxSurge returns the number of machines to surge, or 0 if the deployment strategy does not support surge.
+func (m MachinePoolScope) MaxSurge() (int, error) {
+	if surger, ok := m.deploymentStrategy.(machinepool.Surger); ok {
+		surgeCount, err := surger.Surge(int(m.DesiredReplicas()))
+		if err != nil {
+			return 0, errors.Wrap(err, "failed to calculate surge for the machine pool")
+		}
+
+		return surgeCount, nil
+	}
+
+	return 0, nil
+}
+
+// updateReplicasAndProviderIDs ties the Azure VMSS instance data and the Node status data together to build and update
+// the AzureMachinePool replica count and providerIDList.
+func (m *MachinePoolScope) updateReplicasAndProviderIDs(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolScope.UpdateInstanceStatuses")
+	defer span.End()
+
+	machines, err := m.getMachinePoolMachines(ctx)
 	if err != nil {
-		return errors.Wrap(err, "failed to get node status by provider id")
+		return errors.Wrap(err, "failed to get machine pool machines")
 	}
 
 	var readyReplicas int32
-	instanceStatuses := make([]*infrav1exp.AzureMachinePoolInstanceStatus, len(instances))
-	for i, instance := range instances {
-		instanceStatuses[i] = &infrav1exp.AzureMachinePoolInstanceStatus{
-			ProviderID:        azure.ProviderIDPrefix + instance.ID,
-			InstanceID:        instance.InstanceID,
-			InstanceName:      instance.Name,
-			ProvisioningState: &instance.State,
+	providerIDs := make([]string, len(machines))
+	for i, machine := range machines {
+		if machine.Status.Ready {
+			readyReplicas++
```

Comment:
lots of newly added logic in here, we really need to start adding unit test to these files

Reply:
Couldn't agree more. I was hesitant to add too much more to this PR.